README.rst
Content-negotiation framework for Django
========================================

This project provides a simple and extensible framework for producing views
that content-negotiate in Django.

Prerequisites
-------------

This library depends on Django 1.3, which you can install using your package
manager on recent distributions, or using pip::

    pip install -r requirements.txt

``pip`` is called ``pip-python`` on Fedora. It is generally provided by a
``python-pip`` package.

Using
-----

To define a view, do something like this::

    from django_conneg.views import ContentNegotiatedView

    class IndexView(ContentNegotiatedView):
        def get(self, request):
            context = {
                # Build context here
            }

            # Call render, passing a template name (without file extension)
            return self.render(request, context, 'index')

This will then look for a renderer that can provide a representation that
matches what was asked for in the Accept header.

By default ContentNegotiatedView provides no renderers, so the above snippet
would always return a 405 Not Acceptable to tell the user-agent that it
couldn't provide a response in a suggested format.

To define a renderer on a view, do something like this::

    import json

    from django.http import HttpResponse

    from django_conneg.decorators import renderer

    class JSONView(ContentNegotiatedView):
        @renderer(format='json', mimetypes=('application/json',), name='JSON')
        def render_json(self, request, context, template_name):
            # Very simplistic, and will fail when it encounters 'non-primitives'
            # like Django Model objects, Forms, etc.
            return HttpResponse(json.dumps(context), mimetype='application/json')

.. note::
   ``django-conneg`` already provides a slightly more sophisticated JSONView;
   see below for more information.

You can render to a particular format by calling ``render_to_format()`` on the
view::

    class IndexView(ContentNegotiatedView):
        def get(self, request):
            # ...

            if some_condition:
                return self.render_to_format(request, context, 'index', 'html')
            else:
                return self.render(request, context, 'index')
    

Forcing a particular renderer from the client
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

By default, a client can request a particular set of renderers be tried by
using the ``format`` query or POST parameter::

    GET /some-view/?format=json,yaml

The formats correspond to the ``format`` argument to the ``@renderer``
decorator.

To change the name of the parameter used, override
``_format_override_parameter`` on the view class::

    class MyView(ContentNegotiatedView):
        _format_override_parameter = 'output'


Providing fallback renderers
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Sometimes you might want to provide a response in some format even if the
those in the Accept header can't be honoured. This is useful when providing
error responses in a different format to the client's expected format. To do
this, set the ``_force_fallback_format`` attribute to the name of the format::

    class MyView(ContentNegotiatedView):
        _force_fallback_format = 'html'

If a client doesn't provide an Accept header, then you can specify a default
format with ``_default_format``::

    class MyView(ContentNegotiatedView):
        _default_format = 'html'

Built-in renderer views
~~~~~~~~~~~~~~~~~~~~~~~

``django_conneg`` includes the following built-in renderers in the
``django_conneg.views`` module:

* ``HTMLView`` (renders a ``.html`` template with media type ``text/html``)
* ``TextView`` (renders a ``.txt`` template with media type ``text/plain``)
* ``JSONView`` (coerces the context to JavaScript primitives and returns as ``application/json``)
* ``JSONPView`` (as ``JSONView``, but wraps in a callback and returns as ``application/javascript``)

Using these, you could define a view that renders to both HTML and JSON like this::

    from django_conneg.views import HTMLView

    class IndexView(JSONView, HTMLView):
        def get(self, request):
            # ...
            return self.render(request, context, 'index')

Accessing renderer details
--------------------------

The renderer used to construct a response is exposed as a ``renderer``
attribute on the response object::

    class IndexView(JSONView, HTMLView):
        def get(self, request):
            # ...
            response = self.render(request, context, 'index')
            response['X-Renderer-Format'] = response.renderer.format
            return response 


Renderer priorities
-------------------

Some user-agents might specify various media types with equal levels of
desirability. For example, previous versions of Safari and Chrome `used to send
<http://www.gethifi.com/blog/browser-rest-http-accept-headers#highlighter_222123>`_
an ``Accept`` header like this::

     application/xml,application/xhtml+xml,text/html;q=0.9,
     text/plain;q=0.8,image/png,*/*;q=0.5

Without any additional hints it would be non-deterministic as to whether XML or
XHTML is served.

By passing a ``priority`` argument to the ``@renderer`` decorator you can
specify an ordering of renderers for such ambiguous situations::

     class IndexView(ContentNegotiatedView):
         @renderer(format='xml', mimetypes=('application/xml',), name='XML', priority=0)
         def render_xml(request, context, template_name):
             # ...

         @renderer(format='html', mimetypes=('application/xhtml+xml','text/html), name='HTML', priority=1)
         def render_html(request, context, template_name):
             # ...

As higher-numbered priorities are preferred, this will result in HTML always
being prefered over XML in ambiguous situations.

By default, ``django-conneg``'s built-in renderers have a priority of 0, except
for ``HTMLView`` and ``TextView``, which each have a priority of 1 for the
reason given above.


Improved 40x response handling
------------------------------

Django provides a couple of useful exceptions, ``Http404`` and
``PermissionDenied``, which you may want to use in your application. However,
it's only possible to customise the 404 site-wide (either by providing a
``404.html`` template, or by setting ``handler404`` in your urlconf), and
until Django 1.4 comes out, PermissionDenied will always result in a very
spartan error page.

``django-conneg`` provides an ``ErrorCatchingView`` which you can use as a
mixin to customise the rendering of responses for these error situations::

    from django_conneg.views import HTMLView, ErrorCatchingView

    class IndexView(HTMLView, ErrorCatchingView):
        # ...

You can then customise error responses in one of the following ways:

* overriding the ``conneg/(forbidden|not_found|not_acceptable).(html|txt) templates
* overriding ``error_403``, ``error_404`` or ``error_406`` methods on the view
* overriding the ``error_template_names`` attribute to specify a non-standard template name:

In the latter case, you can do something like::

    import httplib
    from django.util.datastructures import MergeDict
    from django_conneg.views import HTMLView, ErrorCatchingView

    class IndexView(HTMLView, ErrorCatchingView):
        # Provide a view-specific 404 page. Use MergeDict to use django_conneg's
        # defaults for other types of errors.
        error_template_names = MergeDict({httplib.NOT_FOUND: 'foo/404'},
                                         ErrorCatchingView.error_template_names)
        # ...


Running the tests
-----------------

``django-conneg`` has a modest test suite. To run it, head to the root of the
repository and run::

    django-admin test --settings=django_conneg.test_settings --pythonpath=.

If you don't have Django, you'll need to install it as detailed in the
Prerequisites section above.


debian/changelog
python-django-conneg (0.10.1) unstable; urgency=low

  * Use mimetype parameter to HttpResponse if an older version of Django

 -- Alexander Dutton <alexander.dutton@it.ox.ac.uk>  Wed, 17 Dec 2014 11:10:20 +0000

python-django-conneg (0.10.0) unstable; urgency=low

  * Use content_type parameter to HttpResponse

 -- Alexander Dutton <alexander.dutton@it.ox.ac.uk>  Fri, 10 Oct 2014 15:19:07 +0100

python-django-conneg (0.9.5) unstable; urgency=low

  * More HTTP error classes
  * Integer JS timestamps

 -- Alexander Dutton <alexander.dutton@it.ox.ac.uk>  Fri, 7 Mar 2014 22:42:28 +0000

python-django-conneg (0.9.4) unstable; urgency=low

  * Python 3 support
  * format can now be passed from urlconf param
  * BasicAuthMiddleware improvements

 -- Alexander Dutton <alexander.dutton@it.ox.ac.uk>  Thu, 28 Nov 2013 22:23:19 +0000

python-django-conneg (0.9.3) unstable; urgency=low

  * URL in setup.py now points at the right repository.

 -- Alexander Dutton <alexander.dutton@it.ox.ac.uk>  Fri, 23 Nov 2012 09:10:41 +0000

python-django-conneg (0.9.2) unstable; urgency=low

  * Fixed test stage of Debian package (new settings path)
  * Backported to mock < 0.7 (as is on squeeze)

 -- Alexander Dutton <alexander.dutton@it.ox.ac.uk>  Fri, 16 Nov 2012 15:32:47 +0000

python-django-conneg (0.9.1) unstable; urgency=low

  * BasicAuthMiddleware now returns 401 if supplied with invalid credentials,
    and will return 403 for inactive users.
  * Tests for BasicAuthMiddleware

 -- Alexander Dutton <alexander.dutton@it.ox.ac.uk>  Fri, 16 Nov 2012 15:12:45 +0000

python-django-conneg (0.9) unstable; urgency=low

  * More generic handling of HTTP-based exceptions
  * No longer catches and re-raises Exception
  * Datetimes coerced to UTC before JSON serialization

 -- Alexander Dutton <alexander.dutton@oucs.ox.ac.uk>  Fri, 9 Nov 2012 14:16:15 +0000

python-django-conneg (0.8) unstable; urgency=low

  * Tidied some methods into better places and added deprecation warnings.
  * Added test param to renderer decorator to give an indication of whether a
    particular renderer will succeed.
  * renderer decorator now creates callable Renderer objects
  * Lists of renderers can now be combined
  * Some conneg stuff factored out from views into new conneg.Conneg class
  * ErrorCatchingView now default behaviour; provides default templates for
    403 and 404 errors in HTML, JSON and plain text
  * Undocumented attributes view._renderers{,_by_mimetype,_by_format} now
    available as view.conneg.renderers[…]
  * renderer details now placed in context (useful for links to other formats
    when creating templates)

 -- Alexander Dutton <alexander.dutton@oucs.ox.ac.uk>  Fri, 28 Sep 2012 13:08:01 +0100

python-django-conneg (0.7.4) unstable; urgency=low

  * Fixed the raise statement in error_500 so that the original traceback is
    preserved.
  * Allow multiple fallback formats.
  * Added support for HEAD requests.

 -- Alexander Dutton <alexander.dutton@oucs.ox.ac.uk>  Fri, 4 May 2012 14:08:39 +0100

python-django-conneg (0.7.3) unstable; urgency=low

  * Template names in ErrorCatchingView are now configurable.

 -- Alexander Dutton <alexander.dutton@oucs.ox.ac.uk>  Tue, 6 Mar 2012 14:52:12 +0000

python-django-conneg (0.7.2) unstable; urgency=low

  * Exception messages are now passed through to the template by
    ErrorCatchingView and displayed appropriately.

 -- Alexander Dutton <alexander.dutton@oucs.ox.ac.uk>  Tue, 6 Mar 2012 13:56:44 +0000

python-django-conneg (0.7.1) unstable; urgency=low

  * Added ErrorCatchingView for providing content-negotiated responses to
    common error scenarios.
  * JSONView now indents two spaces by default.

 -- Alexander Dutton <alexander.dutton@oucs.ox.ac.uk>  Tue, 6 Mar 2012 11:44:21 +0000

python-django-conneg (0.6) unstable; urgency=low

  * Fixed renderer priority handling

 -- Alexander Dutton <alexander.dutton@oucs.ox.ac.uk>  Thu, 26 Jan 2012 09:31:38 +0000

python-django-conneg (0.5) unstable; urgency=low

  * Initial Release.

 -- Alexander Dutton <alexander.dutton@oucs.ox.ac.uk>  Mon, 19 Dec 2011 14:14:33 +0000


django_conneg/tests.py
import itertools
import unittest

from django.http import HttpResponse
from django.test.simple import DjangoTestSuiteRunner

from . import http, views, decorators

class DatabaselessTestSuiteRunner(DjangoTestSuiteRunner):
    def setup_databases(self, *args, **kwargs): pass
    def teardown_databases(self, *args, **kwargs): pass

class PriorityTestCase(unittest.TestCase):
    mimetypes = ('text/plain', 'application/xml', 'text/html', 'application/json')

    def getRenderer(self, format, mimetypes, name, priority):
        if not isinstance(mimetypes, tuple):
            mimetypes = (mimetypes,)
        def renderer(request, context, template_name):
            return HttpResponse('', mimetype=mimetypes[0])
        renderer.__name__ = 'render_%s' % mimetypes[0].replace('/', '_')
        renderer = decorators.renderer(format=format,
                                       mimetypes=mimetypes,
                                       priority=priority)(renderer)
        return renderer

    def getTestView(self, priorities):
        members = {}
        for i, (mimetype, priority) in enumerate(priorities.items()):
            members['render_%d' % i] = self.getRenderer(str(i), mimetype, str(i), priority)
        TestView = type('TestView',
                        (views.ContentNegotiatedView,),
                        members)
        return TestView

    def testEqualQuality(self):
        accept_header = ', '.join(self.mimetypes)
        accept = http.MediaType.parse_accept_header(accept_header)

        for mimetypes in itertools.permutations(self.mimetypes):
            renderers = tuple(self.getRenderer(str(i), mimetype, str(i), -i) for i, mimetype in enumerate(mimetypes))

            renderers = http.MediaType.resolve(accept, renderers)

            for renderer, mimetype in zip(renderers, mimetypes):
                self.assertEqual(iter(renderer.mimetypes).next(), http.MediaType(mimetype))

    def testEqualQualityView(self):
        accept_header = ', '.join(self.mimetypes)
        accept = http.MediaType.parse_accept_header(accept_header)

        for mimetypes in itertools.permutations(self.mimetypes):
            priorities = dict((mimetype, -i) for i, mimetype in enumerate(mimetypes))
            test_view = self.getTestView(priorities).as_view()
            renderers = http.MediaType.resolve(accept, test_view.conneg.renderers)

            for renderer, mimetype in zip(renderers, mimetypes):
                self.assertEqual(iter(renderer.mimetypes).next(), http.MediaType(mimetype))

    def testPrioritySorting(self):
        for mimetypes in itertools.permutations(self.mimetypes):
            priorities = dict((mimetype, -i) for i, mimetype in enumerate(mimetypes))
            test_view = self.getTestView(priorities).as_view()

            renderer_priorities = [renderer.priority for renderer in test_view.conneg.renderers]
            self.assertEqual(renderer_priorities, sorted(renderer_priorities, reverse=True))



if __name__ == '__main__':
    unittest.main()


django_conneg/tests/urls.py
from django.conf.urls.defaults import patterns, url
from django.conf import settings
from django.http import HttpResponse, HttpResponseRedirect
from django.utils.decorators import method_decorator
from django.contrib.auth.decorators import login_required
from django_conneg.views import HTMLView, JSONView

class OptionalAuthView(HTMLView, JSONView):
    _force_fallback_format = ('html', 'json')
    def get(self, request):
        response = self.render()
        response.is_authenticated = request.user.is_authenticated()
        return response

class LoginRequiredView(HTMLView, JSONView):
    _force_fallback_format = ('html', 'json')
    @method_decorator(login_required)
    def get(self, request):
        response = self.render()
        response.is_authenticated = request.user.is_authenticated()
        return response

urlpatterns = patterns('',
    url(r'^optional-auth/$', OptionalAuthView.as_view()),
    url(r'^login-required/$', LoginRequiredView.as_view()),
)

django_conneg/test_settings.py
import imp

INSTALLED_APPS = (
    'django_conneg',
    'django_conneg.tests',
)

SECRET_KEY = 'test secret key'

# Use django_jenkins if it's installed.
try:
    imp.find_module('django_jenkins')
except ImportError:
    pass
else:
    INSTALLED_APPS += ('django_jenkins',)

DATABASES = {'default': {'ENGINE': 'django.db.backends.sqlite3'}}

LOGIN_URL = '/login/'

ROOT_URLCONF = 'django_conneg.tests.urls'

BASIC_AUTH_ALLOW_HTTP = True

MIDDLEWARE_CLASSES = (
    'django.contrib.sessions.middleware.SessionMiddleware',
    'django.contrib.auth.middleware.AuthenticationMiddleware',
    'django_conneg.support.middleware.BasicAuthMiddleware',
)


django_conneg/decorators.py
from django_conneg.conneg import Renderer

def renderer(format, mimetypes=(), priority=0, name=None, test=None):
    """
    Decorates a view method to say that it renders a particular format and mimetypes.

    Use as:
        @renderer(format="foo")
        def render_foo(self, request, context, template_name): ...
    or
        @renderer(format="foo", mimetypes=("application/x-foo",))
        def render_foo(self, request, context, template_name): ...
    
    The former case will inherit mimetypes from the previous renderer for that
    format in the MRO. Where there isn't one, it will default to the empty
    tuple.

    Takes an optional priority argument to resolve ties between renderers.
    """

    def g(f):
        return Renderer(f, format, mimetypes, priority, name, test)
    return g


django_conneg/views.py
from __future__ import unicode_literals

import datetime
try: # Python < 3
    import httplib as http_client
    import urlparse as urllib_parse
    from urllib import urlencode
    str_types = (unicode, str)
except ImportError: # Python >= 3
    import http.client as http_client
    import urllib.parse as urllib_parse
    from urllib.parse import urlencode
    str_types = (str,)
    unicode = str
import inspect
import itertools
import logging
import sys
import time
import urllib
import warnings

from django.core import exceptions
from django.views.generic import View
from django.utils.decorators import classonlymethod
from django import http
from django.template import RequestContext, TemplateDoesNotExist
from django.shortcuts import render_to_response
from django.utils.cache import patch_vary_headers

from django_conneg.conneg import Conneg
from django_conneg.decorators import renderer
from django_conneg.http import MediaType, HttpError, HttpNotAcceptable
from django_conneg.utils import utc, content_type_arg

logger = logging.getLogger(__name__)

class BaseContentNegotiatedView(View):
    conneg = None
    context = None
    _default_format = None
    _force_fallback_format = None
    _format_override_parameter = 'format'
    _format_url_parameter = 'format'
    _include_renderer_details_in_context = True
    
    template_name = None

    @classonlymethod
    def as_view(cls, **initkwargs):
        view = super(BaseContentNegotiatedView, cls).as_view(**initkwargs)
        view.conneg = Conneg(obj=cls)
        return view

    def dispatch(self, request, *args, **kwargs):
        # This is handy for the view to work out what renderers will
        # be attempted, and to manipulate the list if necessary.
        # Also handy for middleware to check whether the view was a
        # BaseContentNegotiatedView, and which renderers were preferred.
        if self.context is None:
            self.context = {'additional_headers': {}}

        format_url_parameter = kwargs.pop(self._format_url_parameter, None)
        if format_url_parameter:
            self.format_override = [format_url_parameter]
        elif request.REQUEST.get(self._format_override_parameter):
            self.format_override = request.REQUEST[self._format_override_parameter].split(',')
        else:
            self.format_override = None

        self.request = request
        self.args = args
        self.kwargs = kwargs
        self.conneg = Conneg(obj=self)
        self.set_renderers(request)
        return super(BaseContentNegotiatedView, self).dispatch(request, *args, **kwargs)

    def set_renderers(self, request=None, context=None, template_name=None, early=False):
        """
        Makes sure that the renderers attribute on the request is up
        to date. renderers_for_view keeps track of the view that
        is attempting to render the request, so that if the request
        has been delegated to another view we know to recalculate
        the applicable renderers. When called multiple times on the
        same view this will be very low-cost for subsequent calls.
        """
        request, context, template_name = self.get_render_params(request, context, template_name)

        args = (self.conneg, context, template_name,
                self._default_format, self._force_fallback_format, self._format_override_parameter)
        if getattr(request, 'renderers_for_args', None) != args:
            fallback_formats = self._force_fallback_format or ()
            if not isinstance(fallback_formats, (list, tuple)):
                fallback_formats = (fallback_formats,)
            request.renderers = self.conneg.get_renderers(request=request,
                                                          context=context,
                                                          template_name=template_name,
                                                          accept_header=request.META.get('HTTP_ACCEPT'),
                                                          formats=self.format_override,
                                                          default_format=self._default_format,
                                                          fallback_formats=fallback_formats,
                                                          early=early)
            request.renderers_for_view = args
        if self._include_renderer_details_in_context:
            self.context['renderers'] = [self.renderer_for_context(request, r) for r in self.conneg.renderers]
        return request.renderers

    def get_render_params(self, request, context, template_name):
        if not template_name:
            template_name = self.template_name
            if isinstance(template_name, str_types) and template_name.endswith('.html'):
                template_name = template_name[:-5]
        return request or self.request, context or self.context, template_name

    def render(self, request=None, context=None, template_name=None):
        """
        Returns a HttpResponse of the right media type as specified by the
        request.
        
        context can contain status_code and additional_headers members, to set
        the HTTP status code and headers of the request, respectively.
        template_name should lack a file-type suffix (e.g. '.html', as
        renderers will append this as necessary.
        """
        request, context, template_name = self.get_render_params(request, context, template_name)

        self.set_renderers()

        status_code = context.pop('status_code', http_client.OK)
        additional_headers = context.pop('additional_headers', {})

        for renderer in request.renderers:
            response = renderer(request, context, template_name)
            if response is NotImplemented:
                continue
            response.status_code = status_code
            response.renderer = renderer
            break
        else:
            tried_mimetypes = list(itertools.chain(*[r.mimetypes for r in request.renderers]))
            response = self.http_not_acceptable(request, tried_mimetypes)
            response.renderer = None
        for key, value in additional_headers.items():
            response[key] = value

        # We're doing content-negotiation, so tell the user-agent that the
        # response will vary depending on the accept header.
        patch_vary_headers(response, ('Accept',))
        return response

    def http_not_acceptable(self, request, tried_mimetypes, *args, **kwargs):
        response = http.HttpResponse("""\
Your Accept header didn't contain any supported media ranges.

Supported ranges are:

 * %s\n""" % '\n * '.join(sorted('%s (%s; %s)' % (f.name, ", ".join(m.value for m in f.mimetypes), f.format) for f in request.renderers if not any(m in tried_mimetypes for m in f.mimetypes))), **{content_type_arg: "text/plain"})
        response.status_code = http_client.NOT_ACCEPTABLE
        return response

    def head(self, request, *args, **kwargs):
        handle_get = getattr(self, 'get', None)
        if handle_get:
            response = handle_get(request, *args, **kwargs)
            response.content = ''
            return response
        else:
            return self.http_method_not_allowed(request, *args, **kwargs)

    def options(self, request, *args, **kwargs):
        response = http.HttpResponse()
        response['Accept'] = ','.join(m.upper() for m in sorted(self.http_method_names) if hasattr(self, m))
        return response

    @classmethod
    def parse_accept_header(cls, accept):
        warnings.warn("The parse_accept_header method has moved to django_conneg.http.MediaType")
        return MediaType.parse_accept_header(accept)

    def render_to_format(self, request=None, context=None, template_name=None, format=None):
        request, context, template_name = self.get_render_params(request, context, template_name)
        self.set_renderers()

        status_code = context.pop('status_code', http_client.OK)
        additional_headers = context.pop('additional_headers', {})

        for renderer in self.conneg.renderers_by_format.get(format, ()):
            response = renderer(request, context, template_name)
            if response is not NotImplemented:
                break
        else:
            response = self.http_not_acceptable(request, ())
            renderer = None

        response.status_code = status_code
        response.renderer = renderer
        for key, value in additional_headers.items():
            response[key] = value
        return response

    def join_template_name(self, template_name, extension):
        """
        Appends an extension to a template_name or list of template_names.
        """
        if template_name is None:
            return None
        if isinstance(template_name, (list, tuple)):
            return tuple('.'.join([n, extension]) for n in template_name)
        if isinstance(template_name, str_types):
            return '.'.join([template_name, extension])
        raise AssertionError('template_name not of correct type: %r' % type(template_name))

    def renderer_for_context(self, request, renderer):
        return {'name': renderer.name,
                'priority': renderer.priority,
                'mimetypes': [m.value for m in renderer.mimetypes],
                'format': renderer.format,
                'url': self.url_for_format(request, renderer.format)}

    def url_for_format(self, request, format):
        qs = urllib_parse.parse_qs(request.META.get('QUERY_STRING', ''))
        qs['format'] = [format]
        return '?{0}'.format(urlencode(qs, True))



class ContentNegotiatedView(BaseContentNegotiatedView):
    @property
    def error_view(self):
        if not hasattr(self, '_error_view'):
            self._error_view = ErrorView.as_view()
        return self._error_view

    error_template_names = {http_client.NOT_FOUND: ('conneg/not_found', '404'),
                            http_client.FORBIDDEN: ('conneg/forbidden', '403'),
                            http_client.NOT_ACCEPTABLE: ('conneg/not_acceptable',),
                            http_client.BAD_REQUEST: ('conneg/bad_request', '400'),
                            http_client.SERVICE_UNAVAILABLE: ('conneg/service_unavailable', '503'),
                            'default': ('conneg/error',)}

    def dispatch(self, request, *args, **kwargs):
        try:
            return super(ContentNegotiatedView, self).dispatch(request, *args, **kwargs)
        except http.Http404 as e:
            return self.error(request, e, args, kwargs, http_client.NOT_FOUND)
        except exceptions.PermissionDenied as e:
            return self.error(request, e, args, kwargs, http_client.FORBIDDEN)
        except HttpError as e:
            return self.error(request, e, args, kwargs, e.status_code)

    def http_not_acceptable(self, request, tried_mimetypes, *args, **kwargs):
        raise HttpNotAcceptable(tried_mimetypes)

    def error(self, request, exception, args, kwargs, status_code):
        method_name = 'error_%d' % status_code
        method = getattr(self, method_name, None)

        # See if we've got a dedicated handler for this status code
        if callable(method):
            return method(request, exception, *args, **kwargs)

        # Otherwise, if it's an HttpError, try to render it to an
        # appropriate template
        context = {'error': {'status_code': status_code,
                             'status_message': http_client.responses.get(status_code)}}
        if isinstance(exception, HttpError) and exception.args:
            context['error']['message'] = exception.args[0]

        template_names = self.error_template_names.get(status_code,
                                                       self.error_template_names['default'])

        return self.error_view(request, context, template_names)

    def error_406(self, request, exception, *args, **kwargs):
        accept_header_parsed = MediaType.parse_accept_header(request.META.get('HTTP_ACCEPT', ''))
        accept_header_parsed.sort(reverse=True)
        accept_header_parsed = map(unicode, accept_header_parsed)
        context = {'error': {'status_code': http_client.NOT_ACCEPTABLE,
                             'tried_mimetypes': exception.tried_mimetypes,
                             'available_renderers': [self.renderer_for_context(request, r) for r in self.conneg.renderers],
                             'format_parameter_name': self._format_override_parameter,
                             'format_parameter': request.REQUEST.get(self._format_override_parameter),
                             'format_parameter_parsed': request.REQUEST.get(self._format_override_parameter, '').split(','),
                             'accept_header': request.META.get('HTTP_ACCEPT'),
                             'accept_header_parsed': accept_header_parsed}}
        return self.error_view(request, context,
                               self.error_template_names[http_client.NOT_ACCEPTABLE])

# For backwards compatibility
ErrorCatchingView = ContentNegotiatedView

class HTMLView(ContentNegotiatedView):
    _default_format = 'html'

    @renderer(format="html", mimetypes=('text/html', 'application/xhtml+xml'), priority=1, name='HTML')
    def render_html(self, request, context, template_name):
        template_name = self.join_template_name(template_name, 'html')
        if template_name is None:
            return NotImplemented
        try:
            return render_to_response(template_name,
                                      context, context_instance=RequestContext(request),
                                      **{content_type_arg: 'text/html'})
        except TemplateDoesNotExist:
            return NotImplemented

class TextView(ContentNegotiatedView):
    @renderer(format="txt", mimetypes=('text/plain',), priority=1, name='Plain text')
    def render_text(self, request, context, template_name):
        template_name = self.join_template_name(template_name, 'txt')
        if template_name is None:
            return NotImplemented
        try:
            return render_to_response(template_name,
                                      context, context_instance=RequestContext(request),
                                      **{content_type_arg: 'text/plain'})
        except TemplateDoesNotExist:
            return NotImplemented

try:
    import json
except ImportError:
    try:
        import simplejson as json
    except ImportError:
        pass

# Only define if json is available.
if 'json' in locals():
    class JSONView(ContentNegotiatedView):
        _json_indent = 2

        def preprocess_context_for_json(self, context):
            return context

        def simplify_for_json(self, value):
            if inspect.ismethod(getattr(value, 'simplify_for_json', None)):
                return value.simplify_for_json(self.simplify_for_json)
            if isinstance(value, datetime.datetime):
                if value.tzinfo:
                    value = value.astimezone(utc)
                return int(time.mktime(value.timetuple()) * 1000)
            if isinstance(value, (list, tuple)):
                items = []
                for item in value:
                    item = self.simplify_for_json(item)
                    if item is not NotImplemented:
                        items.append(item)
                return items
            if isinstance(value, dict):
                items = {}
                for key, item in value.items():
                    item = self.simplify_for_json(item)
                    if item is not NotImplemented:
                        items[unicode(key)] = item
                return items
            elif type(value) in (int, float, bool):
                return value
            elif type(value) in str_types:
                return unicode(value)
            elif value is None:
                return value
            else:
                logger.warning("Failed to simplify object of type %r", type(value))
                return NotImplemented

        def simplify(self, value):
            warnings.warn("JSONView.simplify() has been renamed to simplify_for_json")
            return self.simplify_for_json(value)

        @renderer(format='json', mimetypes=('application/json',), name='JSON')
        def render_json(self, request, context, template_name):
            context = self.preprocess_context_for_json(context)
            return http.HttpResponse(json.dumps(self.simplify_for_json(context), indent=self._json_indent),
                                     **{content_type_arg: "application/json"})

    class JSONPView(JSONView):
        # The query parameter to look for the callback name
        _default_jsonp_callback_parameter = 'callback'
        # The default callback name if none is provided
        _default_jsonp_callback = 'callback'

        # Overridden to return JSONP if there's a callback parameter
        @renderer(format='json', mimetypes=('application/json',), name='JSON')
        def render_json(self, request, context, template_name):
            if self._default_jsonp_callback_parameter in request.GET:
                return self.render_js(request, context, template_name)
            else:
                return super(JSONPView, self).render_json(request, context, template_name)

        @renderer(format='js', mimetypes=('text/javascript', 'application/javascript'), name='JavaScript (JSONP)')
        def render_js(self, request, context, template_name):
            context = self.preprocess_context_for_json(context)
            callback_name = request.GET.get(self._default_jsonp_callback_parameter,
                                            self._default_jsonp_callback)

            return http.HttpResponse('%s(%s);' % (callback_name, json.dumps(self.simplify_for_json(context), indent=self._json_indent)),
                                     **{content_type_arg: "application/javascript"})

class ErrorView(HTMLView, JSONPView, TextView):
    _force_fallback_format = ('html', 'json')
    def get(self, request, context, template_name):
        self.context.update(context)
        self.template_name = template_name
        self.context['error']['response'] = http_client.responses[context['error']['status_code']]
        self.context['status_code'] = context['error']['status_code']
        return self.render()
    post = delete = put = get
